components/edit_post/edit_post.tsx

```diff
@@ -0,0 +1,494 @@
+    dialogType: typeof React.Component;
+    dialogProps: {
+        post: Post;
+        isRHS?: boolean;
+    };
+};
+
+export type Actions = {
+    addMessageIntoHistory: (message: string) => void;
+    editPost: (input: Partial<Post>) => Promise<Post>;
+    unsetEditingPost: () => void;
+    openModal: (input: OpenModal) => void;
+    scrollPostListToBottom: () => void;
+}
+
+export type Props = {
+    canEditPost?: boolean;
+    canDeletePost?: boolean;
+    readOnlyChannel?: boolean;
+    channelId: string;
+    codeBlockOnCtrlEnter: boolean;
+    ctrlSend: boolean;
+    config: {
+        EnableEmojiPicker?: string;
+        EnableGifPicker?: string;
+    };
+    maxPostSize: number;
+    useChannelMentions: boolean;
+    editingPost: {
+        post: Post | null;
+        postId?: string;
+        refocusId?: string;
+        title?: string;
+        isRHS?: boolean;
+    };
+    actions: Actions;
+};
+
+export type State = {
+    editText: string;
+    caretPosition: number;
+    postError: React.ReactNode;
+    errorClass: string | null;
+    showEmojiPicker: boolean;
+    renderScrollbar: boolean;
+    scrollbarWidth: number;
+    prevShowState: boolean;
+};
+
+const {KeyCodes} = Constants;
+
+const TOP_OFFSET = 0;
+const RIGHT_OFFSET = 10;
+
+const EditPost = ({editingPost, actions, ...rest}: Props): JSX.Element | null => {
+    const [editText, setEditText] = useState<string>(
+        editingPost.post?.message_source || editingPost.post?.message || '',
+    );
+    const [caretPosition, setCaretPosition] = useState<number>(editText.length);
+    const [postError, setPostError] = useState<React.ReactNode | null>(null);
+    const [errorClass, setErrorClass] = useState<string>('');
+    const [showEmojiPicker, setShowEmojiPicker] = useState<boolean>(false);
+    const [renderScrollbar, setRenderScrollbar] = useState<boolean>(false);
+
+    const textboxRef = useRef<TextboxClass>(null);
+    const emojiButtonRef = useRef<HTMLButtonElement>(null);
+    const wrapperRef = useRef<HTMLDivElement>(null);
+
+    const {formatMessage} = useIntl();
+
+    const shouldScroll = (elementRect: DOMRect, containerRect: DOMRect): boolean => {
+        if (!elementRect || !containerRect) {
+            return false;
+        }
+
+        return elementRect.top < containerRect.top || elementRect.bottom > containerRect.bottom;
+    };
+
+    useEffect(() => {
+        const className = `post-list__dynamic${editingPost.isRHS ? '--RHS' : ''}`;
+        const postListElement = document.getElementsByClassName(className)[0];
+
+        const elementCR = wrapperRef?.current?.getBoundingClientRect();
+        const parentCR = postListElement?.getBoundingClientRect();
+
+        if (elementCR && parentCR) {
+            const shouldScrollIntoView = shouldScroll(elementCR, parentCR);
+
+            if (shouldScrollIntoView) {
+                const scrollOptions: boolean|ScrollIntoViewOptions = isFirefox() ? {behavior: 'smooth', block: 'nearest'} : false;
```

Comment:
the method `scrollIntoView` is not very consistent in its behavior. Firefox adds the support for the options, which work way better than just passing the `boolean` for `alignToTop`. The smooth scrolling is way better as well.

For the sake of consistency we could just pass `false` for all browsers, but we would lose a better solution for Firefox. That's something we might need to discuss with UX.

cc/ @matthewbirtch

Reply:
@michelengelen can you clarify what needs to be discussed with UX?

Reply:
@matthewbirtch When scrolling the post that's being edited into view, we're doing it slightly differently on Firefox. If I understand right, we scroll smoothly to the post instead of jumping to it.

I'd be okay if that's the case then. I usually like consistency, but smooth scrolling is much nicer

Reply:
@hmhealey is right about this. Firefox supports the smooth scrolling when using `scrollIntoView`. There are other options (aka npm packagers) out there that support smooth scrolling on all browsers, but I did not want to add a new dependency.

We could absolutely discuss this later on if you would like, but for now I would skip this option for the sake of keeping the PR in scope.

Reply:
Agree that smooth scrolling is an improved experience. I'm ok with leaving it in only for Firefox for now, but it would be best to make it consistent for all browsers. I think that would be okay to do in a follow-up PR though